### PR TITLE
Added annotated_signed_block #936

### PR DESCRIPTION
**Pull request by maslenitsa93**
#936 

# Checklists

Use testnet.

### Check refactored cli_wallet's `get_block` (based on database_api's get_block):

Run following cli-wallet script (**note**: run it after block 4 applied):

```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
get_block 4
```

Result:
```
{
  ...
  "timestamp": ...,
  ...
  "transactions": [{
      ...
      "expiration": ...,
      "operations": [[
          "pow2", ...
        ]
      ],
      "extensions": [],
      "signatures": [
        ...
      ]
    }
  ],
  "block_id": "0000000000000000000000000000000000000000",
  "signing_key": "GLS1111111111111111111111111111111114T1Anm",
  "transaction_ids": []
}
```

### Check new method `get_block_with_virtual_ops`

1. Run this HTML script in browser **before** block 80 applied (**note**: this is a block applying after all hardforks):

```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.send("operation_history", {"method": "get_block_with_virtual_ops", "params":[80]},
    function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result:
```
{
  "previous": "0000000000000000000000000000000000000000",
  "timestamp": "1970-01-01T00:00:00",
  "witness": "",
  "transaction_merkle_root": "0000000000000000000000000000000000000000",
  "extensions": [],
  "witness_signature": "0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
  "transactions": [],
  "block_id": "0000000000000000000000000000000000000000",
  "signing_key": "GLS1111111111111111111111111111111114T1Anm",
  "transaction_ids": []
}
```
2. Now wait for this block and run again. Result:
```
{
  ...
  "timestamp": "2018-08-28T03:35:57",
  ...
  "_virtual_operations": [
    {
      "trx_in_block": 65535,
      "op_in_trx": 0,
      "virtual_op": 1,
      "op": [
        "producer_reward",
        ...
      ]
    }
  ]
}
```

### Check refactored `set_block_applied_callback`

**After all hardforks applied**, run:

```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.send("database_api", {"method": "set_block_applied_callback", "params":["full"]},
    function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result:
```
{
  ...
  "timestamp": ...,
  ...
  "transaction_ids": [],
  "_virtual_operations": [
    {
      "trx_in_block": 65535,
      "op_in_trx": 0,
      "virtual_op": 1,
      "op": [
        "producer_reward",
        ...
      ]
    }
  ]
}
```

### CODE REVIEW

File: libraries/api/block_objects.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        : signed_block(block) {
+    block_id = id();
+    signing_key = signee();
+    transaction_ids.reserve(transactions.size());
+    for (const signed_transaction& tx : transactions) {
+        transaction_ids.push_back(tx.id());
+    }
+}
+
+annotated_signed_block::annotated_signed_block(const signed_block& block, block_operations& ops)
```
Comment (afalaleev) [inline libraries/api/block_objects.cpp]: `(..., const block_operations& ops)`

File: libraries/api/include/golos/api/block_objects.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+using block_operations = std::vector<block_operation>;
+
+struct annotated_signed_block : public signed_block {
+
+    annotated_signed_block();
+
+    annotated_signed_block(const signed_block& block);
+
+    annotated_signed_block(const signed_block& block, block_operations& ops);
```
Comment (afalaleev) [inline libraries/api/include/golos/api/block_objects.hpp]: `(..., const block_operations& ops)`

File: plugins/operation_history/plugin.cpp [hunk 3]
```diff
@@ -120,6 +120,35 @@ namespace golos { namespace plugins { namespace operation_history {
...
         }
 
+        annotated_signed_block get_block_with_virtual_ops(uint32_t block_num) {
+
+            annotated_signed_block result;
+
+            auto sb = database.fetch_block_by_number(block_num);
+            if (!sb.valid()) {
+                return result;
+            } else {
```
Comment (afalaleev) [inline plugins/operation_history/plugin.cpp]: For what `else` ?

File: plugins/operation_history/plugin.cpp [hunk 4]
```diff
@@ -120,6 +120,35 @@ namespace golos { namespace plugins { namespace operation_history {
...
+                result = annotated_signed_block(*sb);
+            }
+
+            const auto& idx = database.get_index<operation_index>().indices().get<by_location>();
+            auto itr = idx.lower_bound(block_num);
+            block_operations ops;
```
Comment (afalaleev) [inline plugins/operation_history/plugin.cpp]: `result._virtual_operations = block_operations();`

File: plugins/operation_history/plugin.cpp [hunk 5]
```diff
@@ -120,6 +120,35 @@ namespace golos { namespace plugins { namespace operation_history {
...
+            for (; itr != idx.end() && itr->block == block_num; ++itr) {
+                if (itr->virtual_op != 0) {
+                    block_operation op;
+                    op.trx_in_block = itr->trx_in_block;
+                    op.op_in_trx = itr->op_in_trx;
+                    op.virtual_op = itr->virtual_op;
+                    op.op = fc::raw::unpack<protocol::operation>(itr->serialized_op);
+                    ops.push_back(op);
```
Comment (afalaleev) [inline plugins/operation_history/plugin.cpp]: `result._virtual_operations.push_back(op);`

File: plugins/operation_history/plugin.cpp [hunk 6]
```diff
@@ -120,6 +120,35 @@ namespace golos { namespace plugins { namespace operation_history {
...
+                }
+            }
+            result._virtual_operations = ops;
```
Comment (afalaleev) [inline plugins/operation_history/plugin.cpp]: Remove line.

**Comment by maslenitsa93** [comment]
Fixing problem. 5 min.

**Comment by maslenitsa93** [comment]
Fixed. Review, please.

**Comment by afalaleev** [comment]
Rebase changes to one commit and change HEAD from golos-v0.19.0 to golos-v0.18.5

**Pull request merged**